app/vmalert/alerting.go

```diff
@@ -143,6 +147,28 @@ func (ar *AlertingRule) ID() uint64 {
 	return ar.RuleID
 }
 
+func (ar *AlertingRule) debugf(at time.Time, a *notifier.Alert, format string, args ...interface{}) {
+	if !ar.Debug {
+		return
+	}
+	prefix := fmt.Sprintf("DEBUG rule %q:%q (%d) at %v: ",
+		ar.GroupName, ar.Name, ar.RuleID, at.Format(time.RFC3339))
+	if a != nil {
+		var labelKeys []string
+		for k := range a.Labels {
+			labelKeys = append(labelKeys, k)
+		}
+		sort.Strings(labelKeys)
+		var labels string
+		for _, l := range labelKeys {
+			labels += fmt.Sprintf("%s=%s,", l, a.Labels[l])
+		}
+		labels = strings.TrimRight(labels, ",")
+		prefix += fmt.Sprintf("alert %d (%s) ", a.ID, labels)
+	}
+	logger.Infof(prefix+format, args...)
```

Comment:
This is incorrect usage of Printf-like functions - the first arg must contain the format string only, which is under full control of the programmer. In this case the first arg contains the `prefix` string, which may contain unexpected format options generated from various args passed to `prefix` string at line 154. For example, if `ar.GroupName` contains `%s`, then it will break formatting at the line 169.

The correct solution - is to always use string literal as the first arg to Printf-like functions, so it can be easily inspected by the reader.

Reply:
Thanks! Please see updated